[Localization] Create swift-def-to-yaml-converter tool

HassanElDesouky opened the pull request:
Since #33422 is becoming a complete miss, @xedin and I discussed another approach which is to leave English text in `.def` files and create a tool which will create a YAML file from the texts of the diagnostic messages in `.def` files. When doing that we won't need `en.yaml` to be in the repository.

What I did was deleting the `/localization` folder... I don't think there is a point of keeping it _now_. Maybe, we can bring it back when we have other translations. 
I also created a unit test for the tool... I'm testing both if the files available in the build folder and if the messages from `en.yaml` matches with the messages from `.def` files.

cc @xedin

xedin commented on tools/swift-localize-diagnostics/swift-localize-diagnostics.cpp:
```diff
@@ -0,0 +1,105 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Create a YAML file from the diagnostic messages texts in `.def` files.
```
This is not a localization tool it's a converter from one format into the other - let's rename it to something like `swift-def-to-yaml-converter`.

xedin commented on CMakeLists.txt:
```diff
@@ -1126,15 +1126,6 @@ if(SWIFT_INCLUDE_TOOLS)
-  # Diagnostics information is only useful for the frontend compiler
-  # anyway, so let's only include it if the compiler is being built,
-  # which at the moment seems like if SWIFT_INCLUDE_TOOLS is defined.
-  add_subdirectory(localization)
```
I think it makes more sense to keep `localization` directory - it would make cmake drive converter and serialization of the en.db. Instead you can remove `localization/diagnostics` directory for now.

HassanElDesouky replied:
So we will have a completely empty `/localization` directory here in the repository?

HassanElDesouky replied:
Or do you mean it will only have the CMake file that serializes `en.yaml`

xedin replied:
It's not going to be empty, it's going to have CMakeLists.txt and you can keep diagnostics directory there.

HassanElDesouky replied:
I think it make sense to remove this `COMMAND` 
```cmake
  COMMAND
    "${SWIFT_NATIVE_SWIFT_TOOLS_PATH}/swift-serialize-diagnostics"
      --input-file-path ${CMAKE_BINARY_DIR}/share/swift/diagnostics/en.yaml
      --output-directory ${CMAKE_BINARY_DIR}/share/swift/diagnostics/
```
from https://github.com/apple/swift/blob/master/localization/CMakeLists.txt.
WDYT?

xedin replied:
It has to stay, build and toolchain should have both yaml and db files available.

xedin commented on tools/swift-localize-diagnostics/swift-localize-diagnostics.cpp:
```diff
@@ -0,0 +1,105 @@
+        finalMsg += msg[j];
+      }
+    }
+    OS << "  msg: \"" << finalMsg << "\"\n\n";
```
I think this should be `\r\n`.

xedin commented on tools/swift-localize-diagnostics/swift-localize-diagnostics.cpp:
```diff
@@ -0,0 +1,105 @@
+
+  std::string MainExecutablePath = llvm::sys::fs::getMainExecutable(
+      argv[0], reinterpret_cast<void *>(&anchorForGetMainExecutable));
+  llvm::SmallString<128> DefaultDiagnosticMessagesDir(MainExecutablePath);
```
Instead of trying to infer it, let's provide the location of the YAML to the tool.

HassanElDesouky committed:
[Localization] Create swift-def-to-yaml-converter tool

HassanElDesouky committed:
[unittests] Add `matchDiagnosticMessagesRandomly`

HassanElDesouky commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -38,6 +45,9 @@ static std::string getDefaultLocalizationPath() {
 }
 
+/// Random number in [0,n)
+unsigned randNum(unsigned n) { return unsigned(rand()) % n; }
```
I figured out that I can use this message since it's already has been used here https://github.com/apple/swift/blob/master/utils/test-clustered-bit-vector/generator.cpp#L12

LucianoPAlmeida commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -0,0 +1,88 @@
+  YAMLLocalizationProducer yaml(EnglishLocalization.str());
+
+  yaml.forEachAvailable([](swift::DiagID id, llvm::StringRef translation) {
+    ASSERT_TRUE(diagnosticMessages[static_cast<uint32_t>(id)] == translation);
```
Why not `ASSERT_EQ`?

HassanElDesouky replied:
no reason, what is the difference?

HassanElDesouky replied:
Ahh, I think since we are comparing strings `ASSERT_EQ` will more useful information when the test fails.

HassanElDesouky replied:
Thank you for pointing that out. :)

LucianoPAlmeida replied:
To me, in almost every test framework the advantage is that is more readable, but also is the message when a test failure happens where assert for equality normally shows something like "value1 is not equal to value2"(which can give you hints on why the test is failing) where assert for true will show "expected true got false" which don't say much.
But for the test fail or pass, it doesn't make any difference :)

LucianoPAlmeida commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -0,0 +1,88 @@
+  });
+}
+
+TEST(DefToYAMLConverterTest, matchDiagnosticMessagesRandomly) {
```
This is more out of curiosity, what such a test covers that is not covered by the sequential test? I'm failing to understand which kind of behavior this check against. Sorry but, I'm just curious :)

xedin replied:
It is useful to test different access patterns because we are treating converter/producer as an unknown. Previous test makes sure that data could be read sequentially, this one is more of a "real world" scenario where data would be read at random positions.

LucianoPAlmeida replied:
Ah makes sense! Thanks @xedin :)

HassanElDesouky committed:
[unittests] Use ASSERT_EQ

HassanElDesouky commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -65,7 +65,8 @@ TEST(DefToYAMLConverterTest, matchDiagnosticMessagesSequentially) {
 
   yaml.forEachAvailable([](swift::DiagID id, llvm::StringRef translation) {
-    ASSERT_TRUE(diagnosticMessages[static_cast<uint32_t>(id)] == translation);
+    llvm::StringRef msg = diagnosticMessages[static_cast<uint32_t>(id)];
+    ASSERT_EQ(msg, translation);
```
I had to store the `msg` in a variable so if `ASSERT_EQ` fails it will show a useful error message like: 
```
      Expected: msg
      Which is: "%0 now requires %select{|no}1 new witness table entry"
To be equal to: translation
      Which is: "something else..."
```

owenv commented on tools/swift-def-to-yaml-converter/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+target_link_libraries(swift-def-to-yaml-converter PRIVATE
+  swiftLocalization
+  swiftAST)
```
I don't think swiftAST is needed so long as you're just #including the .def files, and if that's the case removing it would speed up the build

HassanElDesouky committed:
[tools] Remove swiftAST dependency

xedin commented on tools/swift-def-to-yaml-converter/swift-def-to-yaml-converter.cpp:
```diff
@@ -0,0 +1,94 @@
+  llvm::sys::path::append(LocalizedFilePath, defaultLocaleCode);
+  llvm::sys::path::replace_extension(LocalizedFilePath, ".yaml");
+
+  std::error_code error;
```
Any errors while opening a output file have to be handled, currently `error` is ignored...

HassanElDesouky replied:
Out of curiosity, what are the scenarios in which an error might occur when opening an output file?

xedin replied:
Any I/O issue really, it could be that e.g. disk went read-only right after swiftLocalization was built or there is no space left etc.

HassanElDesouky replied:
Thank you so much!

xedin commented on tools/swift-def-to-yaml-converter/swift-def-to-yaml-converter.cpp:
```diff
@@ -0,0 +1,94 @@
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+#include <cstdlib>
```
Missing imports:
 
- `<string>` for `std::string`
- `<system_error>` for `std::err_code`

xedin commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -0,0 +1,90 @@
+#include "llvm/Support/YAMLTraits.h"
+#include "llvm/Support/raw_ostream.h"
+#include "gtest/gtest.h"
+#include <random>
```
There are some imports missing e.g. `<string>` to reference `std::string`.

xedin commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -0,0 +1,90 @@
+  llvm::SmallString<128> EnglishLocalization(getDefaultLocalizationPath());
+  llvm::sys::path::append(EnglishLocalization, "en");
+  llvm::sys::path::replace_extension(EnglishLocalization, ".yaml");
+  YAMLLocalizationProducer yaml(EnglishLocalization.str());
```
I think this test needs to do following:

1. Create a temporary file;
2. use converter to write to that file;
3. use YAMLLocalizationProducer over file populated in step 2
4. Verify that contents of the YAML match what what in .def (which is already done).

HassanElDesouky replied:
Here's what I tried:

0. I modified `DefToYAMLConverter` to also accept a filename path.
1. I constructed a path to the `DefToYAMLConverter` executable.
2. Created a temp file.
3. Used the YAMLLocalizationProducer to write to the temp file path.

My code:
```cpp
  llvm::SmallString<128> defYAMLConverter(getDefToYAMLConverterPath());
  defYAMLConverter.append(" --output-filename=");

  llvm::SmallString<128> tempFilename;
  std::error_code EC = llvm::sys::fs::createTemporaryFile("en", "yaml", tempFilename);
  ASSERT_FALSE(EC);

  defYAMLConverter.append(tempFilename);
  std::system(tempFilename.c_str());
  YAMLLocalizationProducer yaml(tempFilename.str());

  // Finally, do random or sequential testing... 
```

The problems are:
1. I don't know whether `std::system` is the right method for this. I tried to see if there is anything like it in `LLVM` but couldn't find anything.
2. When I run the code above I get the following error `sh: /var/folders/69/pr6q8w_j1pvcvr_dxdckwpzr0000gn/T/en-49fe0f.yaml: Permission denied`. How can I make it so I have permission to write to the file?

xedin commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -0,0 +1,90 @@
+  llvm::SmallString<128> EnglishLocalization(getDefaultLocalizationPath());
+  llvm::sys::path::append(EnglishLocalization, "en");
+  llvm::sys::path::replace_extension(EnglishLocalization, ".yaml");
+  YAMLLocalizationProducer yaml(EnglishLocalization.str());
```
This test should follow the same sequence as previous one.

xedin commented on localization/CMakeLists.txt:
```diff
@@ -3,6 +3,9 @@ add_custom_target(diagnostic-database)
   COMMAND
     ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/diagnostics/ ${CMAKE_BINARY_DIR}/share/swift/diagnostics/
+  COMMAND
+    "${SWIFT_NATIVE_SWIFT_TOOLS_PATH}/swift-def-to-yaml-converter"
```
Always converting might be too time consuming. Could it be done only if there is no `.yaml` and/or `.def` has been changed between the builds?

HassanElDesouky replied:
I'm not sure how can I do that, I tried using DEPENDS and gave it the path to a .def file and didn't modify it, but it created a new file anyways...

xedin replied:
@compnerd Do you know if there a way to check whether a particular file has been modified `DiagnosticsSema.def` (for example)? @HassanElDesouky How long does it take to convert `DiagnosticsAll.def` into yaml?

HassanElDesouky replied:
an average of 15 to 20 ms.

LucianoPAlmeida replied:
@xedin Maybe `add_custom_command` [`DEPENDS`](https://cmake.org/cmake/help/v3.13/command/add_custom_command.html) can do the trick here like `DEPENDS ${PATH_TO_DEF_FILE}` not really sure but I think that could work to only run if `.def` changes

xedin replied:
`DEPENDS` ensures existence of .def files, it would be more useful to check whether it's changed, but I think it's okay since conversion doesn't take too much time...

LucianoPAlmeida replied:
ahh, I thought that DEPENDS could do it for change files as well, but you right. I was just curious to know if cmake has something built-in for this on custom command, but couldn't find anything yet :(

xedin reviewed (changes requested):
Looks much better, but there is still some way to go, especially where tests are converted.

HassanElDesouky committed:
[tools] [unittests] Add missing include headers

HassanElDesouky committed:
[tools] Handle errors when opening the YAML file

HassanElDesouky committed:
[Localization] Add .gitkeep in diagnostics folder

HassanElDesouky commented on tools/swift-def-to-yaml-converter/swift-def-to-yaml-converter.cpp:
```diff
@@ -0,0 +1,101 @@
+                          llvm::sys::fs::F_None);
+
+  if (OS.has_error() || error) {
+    llvm::errs() << LocalizedFilePath.str() << " does not exist\n";
```
@xedin Do you think there might be a better error message for this, or this will be enough?

xedin replied:
I think it has to say that error has occurred while trying to write into a given file and provide an error code.

HassanElDesouky committed:
[tools] Improve error message and add `output-filename` as an option

HassanElDesouky committed:
[unittests] Make tests call the converter on temp files

HassanElDesouky commented on unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp:
```diff
@@ -60,22 +76,37 @@ TEST(DefToYAMLConverterTest, missingLocalizationFiles) {
+  ASSERT_FALSE(EC);
+  llvm::sys::RemoveFileOnSignal(tempFilename);
+  defYAMLConverter.append(tempFilename);
+  std::system(defYAMLConverter.c_str());
```
@xedin  Is `std::system` is the right method for this? I tried looking for something similar in `llvm` but wasn't able to find anything like it.

xedin replied:
This is not exactly what I meant - converter should be property abstracted which would allow it to be called directly just like `YAMLLocalizationProducer`

xedin replied:
Fixed!

xedin committed:
[Localization] Establish dependency between def-to-yaml converter and diagnostic-database

xedin committed:
[Localization] Extract def-to-yaml conversion logic into `DefToYAMLConverter` to ease testing

xedin committed:
[Localization] Establish a dedicated directory for localization tests

xedin committed:
[Localization] Use testing fixture in DefToYAMLConverter

Makes it easy to setup convertion from def-to-yaml and remove
a temporary file.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain

xedin commented:
Although it's not reflected in the status here are the toolchains:

- https://ci.swift.org/job/swift-PR-toolchain-osx/680/artifact/branch-master/swift-PR-33661-680-osx.tar.gz
- https://ci.swift.org/job/swift-PR-toolchain-Linux/446/artifact/branch-master/swift-PR-33661-446-ubuntu16.04.tar.gz

Both have correct information. I'm going to merge this PR now.
